docs/proposals/config_data.md

```diff
@@ -0,0 +1,315 @@
+is:
+
+1. A `ConfigData` object is created
+2. A pod that consumes the configuration data via environment variables is created
+3. The pod is scheduled onto a node
+4. The kubelet retrieves the `ConfigData` resource(s) referenced by the pod and starts the container
+   processes with the appropriate data in environment variables
+
+### Consuming `ConfigData` in Volumes
+
+Many programs read their configuration from configuration files.  `ConfigData` should be possible
+to consume in a volume.  The rough series of events for consuming `ConfigData` this way
+is:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. The Kubelet creates an instance of the volume plugin and calls its `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+
+### Consuming `ConfigData`  Updates
+
+Any long-running system has configuration that is mutated over time.  Changes made to configuration
+data must be made visible to pods consuming data in volumes so that they can respond to those
+changes.
+
+The `resourceVersion` of the `ConfigData` object will be updated by the API server every time the
+object is modified.  After an update, modifications will be made visible to the consumer container:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+// ConfigData holds configuration data for pods to consume.
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
+  // dot followed by valid DNS_SUBDOMAIN.
+	Data map[string]string `json:"data,omitempty"`
```

Comment:
Another idea I was discussing with @pmorie was to have two sets of data, let's call them A and B, and a field to indicate which of them is current (and the other is previous), and a number between 0 and 999, let's call it the crossover, representing tenths-of-percents. (or maybe 9999, hundredths).  (or maybe 1024 to be a fair modulus).

When a consumer (i.e. the volume) wants to use a ConfigData they generate a random int %1000 (or 1024 or whatever we chose above).  If the random number is >= crossover, use whichever dataset is marked as previous.  If the random number is < crossover, use whichever dataset is marked as current.

This allows a user to push new config data and then increment the crossover (as fast or slow as they need) to transition consumers to the new data.

it's a little complicated, so maybe we don't want to do it just now, but I'd like to see us leave room for it.  Maybe we could put a bit of thought into the REST-ish attributes we want?